### ISSUE TITLE
0.4 test failure: Ambiguous module name ‘Data.HashSet’

### ISSUE THREAD

**Issue by juhp** [GetShopTV/swagger2#15]
See fpco/stackage#1010

```
wagger2-0.4/src/Data/Swagger/Schema/Internal.hs:18:8:
    Ambiguous module name ‘Data.HashSet’:
      it was found in multiple packages:
      unordered-containers-0.2.5.1@unord_7E41cj9RvqSJeZ4jSnbs1a hashmap-1.3.0.1@hashm_622M1jMBjyMIwDJT1SEKMF
Test suite doctest: FAIL
```

Test failed in the Stackage Nightly for swagger2-0.4

**Comment by dmjio** [comment]
`hashmap` is deprecated, use `unordered-containers` if possible

**Comment by fizruk** [comment]
I am not sure why this happens, given that `hashmap` is not in `swagger2` dependencies.

Should we use `PackageImports` to explicitly specify `unordered-containers`?
I think that would fix the problem, but it seems weird to manually disambiguate a module that is not ambiguous when you consider dependencies in `.cabal` file.

**Comment by bergmark** [comment]
You are correct in that you shouldn't need PackageImports, strange error. I also queued swagger2 at http://matrix.hackage.haskell.org/package/swagger2

**Comment by bergmark** [comment]
Sorry I missed that this was the test-suite, so matrix won't help. But it did uncover [other build error](http://matrix.hackage.haskell.org/package/swagger2)

**Comment by fizruk** [comment]
@bergmark I am sorry, but where can I find the error? I see no logs there.

**Comment by bergmark** [comment]
By clicking the red (failure) cells

* http://matrix.hackage.haskell.org/package/swagger2#GHC-7.10/swagger2-0.4
* http://matrix.hackage.haskell.org/package/swagger2#GHC-7.8/swagger2-0.4
* http://matrix.hackage.haskell.org/package/swagger2#GHC-7.6/swagger2-0.4

**Comment by fizruk** [comment]
The GHC 7.10 build error is weird because it refers to `deriveJSON`'s generated code.
I see that it uses latest `aeson-0.10.0.0`, so I can probably fix it by setting an upper bound on `aeson`.

@bergmark how can I improve my Travis builds to catch this kind of errors before the release?

**Comment by bergmark** [comment]
I use https://github.com/hvr/multi-ghc-travis to test multiple GHC versions. Matrix essentially just does `cabal install swagger2` so since travis starts with a clean sandbox it should give the same results.

**Comment by fizruk** [comment]
I hoped for a `stack`-based version of `multi-ghc-travis` :)

Okay, will switch back to `cabal-install` then.

**Comment by fizruk** [comment]
@bergmark what should I do about tests failure? Am I correct that the problem was introduced by [this `swagger2.cabal` diff v0.3–v0.4](http://hdiff.luite.com/cgit/swagger2/diff/swagger2.cabal?id=0.4&id2=0.3)?

**Comment by bergmark** [comment]
Yes that seems correct. From the littleI know about doctest it sounds plausible that it doesn't use packages dependencies properly and instead uses the whole packagedb. If so, a PackageImport would solve this. But it may be wise to ask the doctest developers as well.

**Comment by fizruk** [comment]
The GHC 7.10 build failure in matrix is due to `aeson-0.10.0.0` bug (see bos/aeson#293).
I am surprised it hasn't been released yet!

I have added an upper version bound for `aeson` in 9b3470d2723ffefea287063d74f4331c4c26e629 until the fix is released.

**Comment by fizruk** [comment]
Closing this issue since both build problems should be fixed in [v0.4.1](http://hackage.haskell.org/package/swagger2-0.4.1).

**Comment by juhp** [comment]
Thanks!
